documentation/explanation/design-decisions.md

```diff
@@ -0,0 +1,325 @@
+    run &OrderValidator.validate/2
+  end
+  
+  step :process_payment do
+    argument :order, result(:validate_order)
+    argument :payment_info, input(:payment_info)
+    run &PaymentProcessor.charge/2
+    compensate &PaymentProcessor.refund/4
+  end
+end
+```
+
+**Builder API Strengths - Dynamic Workflows:**
+```elixir
+def build_dynamic_workflow(config) do
+  reactor = Reactor.Builder.new()
+  
+  # Add steps based on runtime configuration
+  Enum.reduce(config.processing_steps, reactor, fn step_config, reactor ->
+    {:ok, reactor} = Reactor.Builder.add_step(reactor, step_config.name, step_config.impl)
+    reactor
+  end)
+end
+```
+
+### When to Use Each
+
+**Use DSL When:**
+- Workflow structure is known at compile time
+- You want declarative, self-documenting code
+- Static analysis and validation is important
+
+**Use Builder API When:**
+- Workflow structure depends on runtime data
+- Building workflows programmatically
+- Integrating with code generation tools
+
+**Both approaches produce identical runtime structures and can be composed together.**
+
+## Concurrency by Default
+
+### The Decision
+
+Steps run asynchronously by default (`async?: true`) with sophisticated concurrency management and deadlock prevention.
+
+### Why Async by Default?
+
+**Modern Computing Reality:**
+- Multi-core processors are standard
+- I/O operations dominate many workflows
+- Network latency is orders of magnitude higher than CPU operations
+
+**Performance Benefits:**
+```elixir
+# Without concurrency: 300ms total
+fetch_user(100ms) → fetch_preferences(100ms) → fetch_history(100ms)
+
+# With concurrency: ~100ms total  
+fetch_user(100ms) ┐
+fetch_preferences(100ms) ├→ merge_results
+fetch_history(100ms) ┘
+```
+
+### Sophisticated Concurrency Management
+
+**Deadlock Prevention:**
+```elixir
+# Automatic handling of nested reactor scenarios
+defmodule ParentReactor do
+  use Reactor
+
+  compose :process_batch, ChildReactor do
+    # When concurrency pool exhausted, composed reactor runs synchronously
+    # Automatically detects and prevents deadlock
+  end
+end
+
+# Run with shared concurrency pool
+Reactor.run(ParentReactor, %{}, max_concurrency: 8)
+```
+
+**Resource Management:**
+- Shared concurrency pools across reactor hierarchies
+- Automatic resource acquisition and release
+- Process monitoring and cleanup
+
+### Design Philosophy
+
+**Performance by Default:**
+- Optimal performance without developer intervention
+- Scales automatically with system capabilities
+- Simple mental model: "declare dependencies, get concurrency"
+
+**Safety and Reliability:**
+- Automatic deadlock detection and prevention
+- Resource limits prevent system exhaustion
+- Explicit opt-out for special cases (`async?: false`)
+
+## Integration with Spark/Ash Ecosystem
+
+### The Decision
+
+Reactor is built on the Spark framework and designed to integrate seamlessly with the Ash ecosystem, rather than being a standalone workflow engine.
```

Comment:
```suggestion
Reactor is built on the Spark library and designed to integrate seamlessly with the Ash ecosystem, rather than being a standalone workflow engine.
```

Reply:
Also it is standalone isn't it?